perf/tools/setup_env.sh

```diff
@@ -0,0 +1,134 @@
+#!/bin/sh
+
+# The script setting up a Linux operating system before running
+# Tarantool benchmarks.
+# See https://github.com/tarantool/tarantool/wiki/Benchmarking.
+
+set -eu
+
+uid=$(id -u)
+if [ "$uid" -ne 0 ]
+  then echo "Please run as root."
+  exit 1
+fi
+
+###
+# Helpers.
+###
+
+cpu_vendor="unknown"
+cpuinfo_vendor=$(awk '/vendor_id/{ print $3; exit }' < /proc/cpuinfo)
+if [ "$cpuinfo_vendor" = "GenuineIntel" ]; then
+   cpu_vendor="intel"
+elif [ "$cpuinfo_vendor" = "AuthenticAMD" ]; then
+   cpu_vendor="amd"
+else
+   echo "Unknown CPU vendor '$cpuinfo_vendor'"
+   exit 1
+fi
+
+###
+# Kernel command line parameters.
+###
+
+echo -n "Disable AMD SMT or Intel Hyperthreading "
+sysfs_path="/sys/devices/system/cpu/smt/active"
+if [ -f "$sysfs_path" ]; then
+  is_set=$(cat $sysfs_path)
+  [ "$is_set" = 1 ] && echo "SUCCESS" || echo "FAILED (hint: set 'nosmt' kernel parameter)"
+else
+  echo "SKIPPED"
+fi
+
+echo -n "Isolate CPUs for benchmarking "
```

Comment:
Minor: Maybe it is worth it to introduce the function that takes the path and hint message as arguments (for options `nosmt`, `isolcpus`, `irqaffinity`, `nohz_full`). Also, it will be nice to sort entries alphabetically.
Feel free to ignore.

Reply:
Introduced two functions: setting to a value and check a value for non-zero:
```diff
diff --git a/perf/tools/setup_env.sh b/perf/tools/setup_env.sh
index b29fcad3dc..4be7f7bfa9 100755
--- a/perf/tools/setup_env.sh
+++ b/perf/tools/setup_env.sh
@@ -31,6 +31,33 @@ FAILURE_MSG="WARNING"
 SUCCESS_MSG="CHECKED"
 SKIPPED_MSG="SKIPPED"
 
+set_kernel_setting() {
+  desc_msg="$1"
+  file_path="$2"
+  value="$3"
+
+  echo -n "$desc_msg "
+  if [ -f "$file_path" ]; then
+    sh -c "echo $value > $file_path" && echo $SUCCESS_MSG || echo $FAILURE_MSG
+  else
+    echo $SKIPPED_MSG
+  fi
+}
+
+kernel_setting_is_nonzero() {
+  desc_msg="$1"
+  file_path="$2"
+  hint_msg="$3"
+
+  echo -n "$desc_msg "
+  if [ -f "$file_path" ]; then
+    value=$(cat "$file_path")
+    [ -n "$value" ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: $hint_msg)"
+  else
+    echo $SKIPPED_MSG
+  fi
+}
+
 ###
 # Kernel command line parameters.
 ###
@@ -41,91 +68,50 @@ if [ -f "$sysfs_path" ]; then
   is_set=$(cat $sysfs_path)
   [ "$is_set" = 1 ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: set 'nosmt' kernel parameter)"
 else
-  echo "SKIPPED"
+  echo $SKIPPED_MSG
 fi
 
-echo -n "Isolate CPUs for benchmarking "
+desc_msg="Isolate CPUs for benchmarking"
 sysfs_path="/sys/devices/system/cpu/isolated"
-if [ -f "$sysfs_path" ]; then
-  isolcpus=$(cat "$sysfs_path")
-  [ -n "$isolcpus" ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: set 'isolcpus' kernel parameter)"
-else
-  echo "SKIPPED"
-fi
+hint_msg="set 'isolcpus' kernel parameter"
+kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
 
-echo -n "Offload interrupts from the isolated CPUs "
+desc_msg="Offload interrupts from the isolated CPUs"
 sysfs_path="/proc/irq/default_smp_affinity"
-if [ -f "$sysfs_path" ]; then
-  smp_affinity=$(cat "$sysfs_path")
-  [ -n "$smp_affinity" ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: set 'irqaffinity' kernel parameter)"
-else
-  echo "SKIPPED"
-fi
+hint_msg="set 'irqaffinity' kernel parameter"
+kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
 
-echo -n "Disable scheduling on single-task isolated CPUs "
+desc_msg="Disable scheduling on single-task isolated CPUs"
 sysfs_path="/sys/devices/system/cpu/nohz_full"
-if [ -f "$sysfs_path" ]; then
-  nohz=$(cat "$sysfs_path")
-  [ -n "$nohz" ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: set 'nohz_full' kernel parameter)"
-else
-  echo "FAILED (hint: enable CONFIG_NO_HZ_FULL in a kernel config)"
-fi
+hint_msg="set 'nohz_full' kernel parameter"
+kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
 
-echo -n "Disable transparent_hugepage "
+desc_msg="Disable transparent huge pages"
 sysfs_path="/sys/kernel/mm/transparent_hugepage/enabled"
-if [ -f "$sysfs_path" ]; then
-  sh -c "echo never > $sysfs_path"
-  echo $SUCCESS_MSG
-else
-  echo $SKIPPED_MSG
-fi
+set_kernel_setting "$desc_msg" "$sysfs_path" "never"
 
-echo -n "Disable transparent_hugepage "
+desc_msg="Disable direct compaction of transparent huge pages "
 sysfs_path="/sys/kernel/mm/transparent_hugepage/defrag"
-if [ -f "$sysfs_path" ]; then
-  sh -c "echo never > $sysfs_path"
-  echo $SUCCESS_MSG
-else
-  echo $SKIPPED_MSG
-fi
+set_kernel_setting "$desc_msg" "$sysfs_path" "never"
 
 ###
 # System tuning.
 ###
 
-echo -n "Disable TurboBoost "
+desc_msg="Disable TurboBoost"
 if [ "$cpu_vendor" = "amd" ]; then
   sysfs_path="/sys/devices/system/cpu/cpufreq/boost"
-  if [ -f "$sysfs_path" ]; then
-    sh -c "echo 0 > $sysfs_path"
-    echo $SUCCESS_MSG
-  else
-    echo $SKIPPED_MSG
-  fi
+  value=0
 elif [ "$cpu_vendor" = "intel" ]; then
   sysfs_path="/sys/devices/system/cpu/intel_pstate/no_turbo"
-  if [ -f "$sysfs_path" ]; then
-    sh -c "echo 1 > $sysfs_path"
-    echo $SUCCESS_MSG
-  else
-    echo $SKIPPED_MSG
-  fi
+  value=1
 fi
+set_kernel_setting "$desc_msg" "$sysfs_path" "$value"
 
-echo -n "Stabilize the CPU frequency "
+desc_msg="Stabilize the CPU frequency"
 sysfs_path="/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor"
-if [ -f "$sysfs_path" ]; then
-  sh -c 'echo performance | tee $sysfs_path'
-  echo $SUCCESS_MSG
-else
-  echo $SKIPPED_MSG
-fi
+set_kernel_setting "$desc_msg" "$sysfs_path" "performance"
 
-echo -n "Workload-specific: drop disk caches "
+desc_msg="Workload-specific: drop disk caches"
 proc_path="/proc/sys/vm/drop_caches"
-if [ -f "$proc_path" ]; then
-  sh -c "echo 1 > $proc_path"
-  echo $SUCCESS_MSG
-else
-  echo $SKIPPED_MSG
-fi
+set_kernel_setting "$desc_msg" "$proc_path" 1
```

Reply:
Maybe it is easier for reading to give arguments like the following (same for `set_kernel_setting()`):
```sh
kernel_setting_is_nonzero \
	"Isolate CPUs for benchmarking" \
	"/sys/devices/system/cpu/isolated" \
	"set 'isolcpus' kernel parameter"

kernel_setting_is_nonzero \
	"Offload interrupts from the isolated CPUs" \
	"/proc/irq/default_smp_affinity" \
	"set 'irqaffinity' kernel parameter"

kernel_setting_is_nonzero \
	"Disable scheduling on single-task isolated CPUs" \
	"/sys/devices/system/cpu/nohz_full" \
	"set 'nohz_full' kernel parameter"
```

What do you think?

Reply:
Good idea, updated.

Reply:
```diff
diff --git a/perf/tools/setup_env.sh b/perf/tools/setup_env.sh
index 4be7f7bfa9..792a228ea3 100755
--- a/perf/tools/setup_env.sh
+++ b/perf/tools/setup_env.sh
@@ -66,39 +66,41 @@ echo -n "Disable AMD SMT or Intel Hyperthreading "
 sysfs_path="/sys/devices/system/cpu/smt/active"
 if [ -f "$sysfs_path" ]; then
   is_set=$(cat $sysfs_path)
-  [ "$is_set" = 1 ] && echo $SUCCESS_MSG || echo "$FAILURE_MSG (hint: set 'nosmt' kernel parameter)"
+  msg="$FAILURE_MSG (hint: set 'nosmt' kernel parameter)"
+  [ "$is_set" = 1 ] && echo $SUCCESS_MSG || echo "$msg"
 else
   echo $SKIPPED_MSG
 fi
 
-desc_msg="Isolate CPUs for benchmarking"
-sysfs_path="/sys/devices/system/cpu/isolated"
-hint_msg="set 'isolcpus' kernel parameter"
-kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
+kernel_setting_is_nonzero \
+  "Isolate CPUs for benchmarking" \
+  "/sys/devices/system/cpu/isolated" \
+  "set 'isolcpus' kernel parameter"
 
-desc_msg="Offload interrupts from the isolated CPUs"
-sysfs_path="/proc/irq/default_smp_affinity"
-hint_msg="set 'irqaffinity' kernel parameter"
-kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
+kernel_setting_is_nonzero \
+  "Offload interrupts from the isolated CPUs" \
+  "/proc/irq/default_smp_affinity" \
+  "set 'irqaffinity' kernel parameter"
 
-desc_msg="Disable scheduling on single-task isolated CPUs"
-sysfs_path="/sys/devices/system/cpu/nohz_full"
-hint_msg="set 'nohz_full' kernel parameter"
-kernel_setting_is_nonzero "$desc_msg" "$sysfs_path" "$hint_msg"
+kernel_setting_is_nonzero \
+  "Disable scheduling on single-task isolated CPUs" \
+  "/sys/devices/system/cpu/nohz_full" \
+  "set 'nohz_full' kernel parameter"
 
-desc_msg="Disable transparent huge pages"
-sysfs_path="/sys/kernel/mm/transparent_hugepage/enabled"
-set_kernel_setting "$desc_msg" "$sysfs_path" "never"
+set_kernel_setting \
+  "Disable transparent huge pages" \
+  "/sys/kernel/mm/transparent_hugepage/enabled" \
+  "never"
 
-desc_msg="Disable direct compaction of transparent huge pages "
-sysfs_path="/sys/kernel/mm/transparent_hugepage/defrag"
-set_kernel_setting "$desc_msg" "$sysfs_path" "never"
+set_kernel_setting \
+  "Disable direct compaction of transparent huge pages" \
+  "/sys/kernel/mm/transparent_hugepage/defrag" \
+  "never"
 
 ###
 # System tuning.
 ###
 
-desc_msg="Disable TurboBoost"
 if [ "$cpu_vendor" = "amd" ]; then
   sysfs_path="/sys/devices/system/cpu/cpufreq/boost"
   value=0
@@ -106,12 +108,17 @@ elif [ "$cpu_vendor" = "intel" ]; then
   sysfs_path="/sys/devices/system/cpu/intel_pstate/no_turbo"
   value=1
 fi
-set_kernel_setting "$desc_msg" "$sysfs_path" "$value"
-
-desc_msg="Stabilize the CPU frequency"
-sysfs_path="/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor"
-set_kernel_setting "$desc_msg" "$sysfs_path" "performance"
-
-desc_msg="Workload-specific: drop disk caches"
-proc_path="/proc/sys/vm/drop_caches"
-set_kernel_setting "$desc_msg" "$proc_path" 1
+set_kernel_setting \
+  "Disable TurboBoost" \
+  "$sysfs_path" \
+  "$value"
+
+set_kernel_setting \
+  "Stabilize the CPU frequency" \
+  "/sys/devices/system/cpu/cpu*/cpufreq/scaling_governor" \
+  "performance"
+
+set_kernel_setting \
+  "Workload-specific: drop disk caches" \
+  "/proc/sys/vm/drop_caches" \
+  "1"
```

Reply:
>  Also, it will be nice to sort entries alphabetically.

The order is the same as in benchmarking page. Do you want to have another order in the script?

Reply:
I don't insist, feel free to ignore.

Reply:
ignored